src/ripple/nodestore/impl/DeterministicShard.h

```diff
@@ -0,0 +1,153 @@
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_NODESTORE_DETERMINISTICSHARD_H_INCLUDED
+#define RIPPLE_NODESTORE_DETERMINISTICSHARD_H_INCLUDED
+
+#include <ripple/nodestore/DatabaseShard.h>
+#include <ripple/nodestore/DummyScheduler.h>
+#include <nudb/nudb.hpp>
+#include <set>
+
+// TODO: uncomment after NuDB PR will be approved
+//#define NUDB_CTX_HAS_FLUSH
+
+namespace ripple {
+namespace NodeStore {
+
+/** DeterministicShard class.
+ *
+ * 1. The init() method creates temporary folder tempdir_,
+ *    and the deterministic shard is initialized in that folder.
+ * 2. The store() method adds object to memory pool.
+ * 3. The flush() method stores all objects from memory pool to the shard
+ *    located in tempdir_ in sorted order.
+ * 4. The close(true) method finalizes the shard and moves it from tempdir_
+ *    temporary folder to finaldir_ permanent folder,
+ *    deleting old (non-deterministic) shard located in finaldir_.
+ */
+class DeterministicShard
+{
+    constexpr static std::uint32_t maxMemObjsDef = 16384u;
+    constexpr static std::uint32_t maxMemObjsTest = 16u;
+
+public:
+    using nodeptr = std::shared_ptr<NodeObject>;
+
+    DeterministicShard(DeterministicShard const&) = delete;
+    DeterministicShard&
+    operator=(DeterministicShard const&) = delete;
+
+    /** Creates the object for shard database
+     *
+     * @param app Application object
+     * @param db Shard Database which deterministic shard belongs to
+     * @param index Index of the shard
+     * @param lastHash Hash of last ledger in the shard
+     * @param j Journal to logging
+     */
+    DeterministicShard(
+        Application& app,
+        DatabaseShard const& db,
+        std::uint32_t index,
+        uint256 const& lastHash,
+        beast::Journal j);
+
+    ~DeterministicShard();
+
+    /** Initializes the deterministic shard.
+     *
+     * @return true is success, false if errored
+     */
+    bool
+    init();
+
+    /** Finalizes and closes the shard.
+     *
+     * @param cancel True if reject the shard and delete all files,
+     *               false if finalize the shard and store them
+     */
+    void
+    close(bool cancel = false);
+
+    /** Store the object into memory pool
+     *
+     * @param nobj Object to store.
+     */
+    void
+    store(nodeptr const& nobj);
+
+    /** Flush all objects from memory pool to shard
+     */
+    void
+    flush();
+
+private:
+    // Count hash of shard parameters: lashHash, firstSeq, lastSeq, index
+    uint160
+    hash(const uint256& lastHash) const;
+
+    // Get n-th 64-bit portion of shard parameters's hash
+    std::uint64_t
+    digest(int n) const;
+
+    // If database inited
+    bool inited_;
+
+    // Application reference
+    Application& app_;
+
+    // Shard database
+    DatabaseShard const& db_;
```

Comment:
Needs a comment explaining why the lifetime of `db_` will always outlive this object (Also, note to myself to audit this code to look at the lifetime).

Reply:
Done. db_ excluded.